File: cusum_fault_detection.m
function fault_index = cusum_fault_detection(RMS_signal)

mu0 = mean(RMS_signal(1:20));
k   = 0.01;

CUSUM = zeros(size(RMS_signal));

for i = 2:length(RMS_signal)
    CUSUM(i) = max(0, CUSUM(i-1) + RMS_signal(i) - mu0 - k);
end

threshold   = 0.5;
fault_index = find(CUSUM > threshold, 1);

File: cusum_fault_detection.py
# ============================================================
# cusum_fault_detection.py
# CUSUM-based fault detection on RMS signal
# ============================================================

import numpy as np


def cusum_fault_detection(RMS_signal):

    mu0   = np.mean(RMS_signal[:20])   # baseline from first 20 windows
    k     = 0.01
    CUSUM = np.zeros(len(RMS_signal))

    for i in range(1, len(RMS_signal)):
        CUSUM[i] = max(0, CUSUM[i-1] + RMS_signal[i] - mu0 - k)

    threshold = 0.5
    indices   = np.where(CUSUM > threshold)[0]   # find all indices above threshold

    fault_index = indices[0] if len(indices) > 0 else None

    return fault_index